py/src/nupic/encoders/date.py

```diff
@@ -0,0 +1,316 @@
+
+    self.holidayEncoder = None
+    if holiday != 0:
+      p = ScalarEncoderParameters()
+      # A "continuous" binary value. = 1 on the holiday itself and smooth ramp
+      # 0->1 on the day before the holiday and 1->0 on the day after the
+      # holiday.
+      p.minimum    = 0
+      p.maximum    = 1
+      p.radius     = 1
+      p.activeBits = holiday
+      self.holidayEncoder = ScalarEncoder(p)
+      self.size += self.holidayEncoder.size
+
+      for h in holidays:
+        if not (hasattr(h, "__getitem__") or len(h) not in [2,3]):
+          raise ValueError("Holidays must be an iterable of length 2 or 3")
+      self.holidays = holidays
+
+    self.timeOfDayEncoder = None
+    if timeOfDay != 0:
+      p = ScalarEncoderParameters()
+      p.minimum  = 0
+      p.maximum  = 24
+      p.periodic = True
+      # Value is time of day in hours
+      # Radius = 4 hours, e.g. morning, afternoon, evening, early night,  late
+      # night, etc.
+      try:
+        activeBits, radius = timeOfDay
+      except TypeError:
+        p.activeBits = timeOfDay
+        p.radius     = 4
+      else:
+        p.activeBits = timeOfDay[0]
+        p.radius     = timeOfDay[1]
+
+      self.timeOfDayEncoder = ScalarEncoder(p)
+      self.size += self.timeOfDayEncoder.size
+
+    self.dimensions = (self.size,)
+    assert(self.size > 0)
+
+  def reset(self):
+    """ Does nothing, DateEncoder holds no state. """
+    pass
+
+  def encode(self, inp, output=None):
+    """
+    Argument inp: (datetime) representing the time being encoded
+    """
+    if output is None:
+      output = SDR(self.dimensions)
+    else:
+      assert( isinstance(output, SDR) )
+      assert( output.dimensions == self.dimensions )
+
+    if inp is None or (isinstance(inp, float) and math.isnan(inp)):
+      output.zero()
+      return output
+
+    elif not isinstance(inp, datetime.datetime):
+      raise ValueError("Input is type %s, expected datetime. Value: %s" % (
+          type(inp), str(inp)))
+
+    # -------------------------------------------------------------------------
+    # Encode each sub-field
+    sdrs      = []
+    timetuple = inp.timetuple()
+    timeOfDay = timetuple.tm_hour + float(timetuple.tm_min)/60.0
+
+    if self.seasonEncoder is not None:
+      # Number the days starting at zero, intead of 1 like the datetime does.
+      dayOfYear = timetuple.tm_yday - 1
+      # dayOfYear -= self.seasonEncoder.parameters.radius / 2. # Round towards the middle of the season.
+      sdrs.append( self.seasonEncoder.encode(dayOfYear) )
+
+    if self.dayOfWeekEncoder is not None:
+      dayOfWeek = timetuple.tm_wday + (timeOfDay) / 24.0
+      dayOfWeek -= .5 # Round towards noon, not midnight
+      sdrs.append( self.dayOfWeekEncoder.encode(dayOfWeek) )
+
+    if self.weekendEncoder is not None:
+      # saturday, sunday or friday evening
+      if (timetuple.tm_wday == 6 or timetuple.tm_wday == 5 or
+         (timetuple.tm_wday == 4 and timeOfDay > 18)):
```

Comment:
This is the logic for weekend encoding.  It classifies Friday after 6pm as weekend.  Custom days can not do this; custom Days works with whole days only.

This encoder has bells and whistles.  This should not cause problems because it's disabled by default.